aten/src/ATen/native/SpectralOps.cpp

```diff
@@ -5,33 +5,152 @@
 #endif
 
 #include "ATen/ATen.h"
+#include "ATen/Config.h"
 #include "ATen/NativeFunctions.h"
-#include "ATen/WrapDimUtils.h"
-#include "ATen/ExpandUtils.h"
+#include "ATen/native/SpectralOpsUtils.h"
 
 #include <algorithm>
-#include <functional>
-#include <numeric>
 #include <vector>
-
+#include <cmath>
 
 namespace at { namespace native {
 
-// Since real-to-complex satisfy the Hermitian symmetry, i.e.,
-// X[m, \omega] = X[m, N - \omega]*. We return only the first floor(N / 2) + 1
-// values by default. This is also the assumption in libraries including cuFFT.
-static inline int64_t infer_ft_complex_length(int64_t real_length) {
-  return (real_length >> 1) + 1;
+static inline Tensor _fft(const Tensor &self, const int64_t signal_ndim,
+           const bool complex_input, const bool complex_output,
+           const bool inverse, IntList signal_sizes, const bool normalized,
+           const bool onesided) {
+
+  if (signal_ndim < 1 || signal_ndim > 3) {
+    std::ostringstream ss;
+    ss << "Expected signal_ndim to be 1, 2, or 3, but got signal_ndim="
+       << signal_ndim;
+    throw std::runtime_error(ss.str());
+  }
+  if (!at::isFloatingType(self.type().scalarType())) {
+    std::ostringstream ss;
+    ss << "Expected an input tensor of floating types, but got input "
+       << self.type() << self.sizes();
+    throw std::runtime_error(ss.str());
+  }
+
+  auto signal_tensor_ndim = signal_ndim + (int) complex_input;  // add complex dim
+  if (self.dim() != signal_tensor_ndim && self.dim() != signal_tensor_ndim + 1) {
+    std::ostringstream ss;
+    ss << "Given signal_ndim=" << signal_ndim << ", expected an input tensor "
+       << "of " << signal_tensor_ndim << "D or " << signal_tensor_ndim + 1
+       << "D (batch mode)";
+    if (complex_input) {
+      ss << " (complex input adds an extra dimension)";
+    }
+    ss << ", but got input " << self.type() << self.sizes();
+    throw std::runtime_error(ss.str());
+  }
+  bool is_batched = self.dim() == signal_tensor_ndim + 1;
+
+  Tensor input = self;
+
+  if (!is_batched) {
+    input = input.unsqueeze(0);
+  }
+
+  if (complex_input) {
+    if (input.size(signal_ndim + 1) != 2) {
+      std::ostringstream ss;
+      ss << "Expected an input tensor with a last dimension of size 2 "
+         << "representing real + imaginary components, but got input "
+         << self.type() << self.sizes();
+      throw std::runtime_error(ss.str());
+    }
+  }
+
+  // build signal_sizes and output_size
+  if (signal_sizes.size() > 0 && (int64_t) signal_sizes.size() != signal_ndim) {
+    std::ostringstream ss;
+    ss << "Expected signal_sizes to be empty (default) or of signal_ndim="
+       << signal_ndim << "D, but got signal_sizes=" << signal_sizes;
+    throw std::runtime_error(ss.str());
+  }
+  std::vector<int64_t> output_sizes(signal_ndim + 1 + (int) complex_output);
+  output_sizes[0] = input.size(0);
+  std::vector<int64_t> checked_signal_sizes(signal_ndim);
+  int64_t input_size;
```

Comment:
You can push the variable declaration inside the loop, because it is not carried.